Screen 2 invest withdraw

BartDeCaluwe opened the pull request:
This PR will add:

- error and success messages
- invest-withdraw page where a user can deposit or withdraw from the liquidity pool

BartDeCaluwe committed:
feat: wrap input with Form.Item component

AntD Form doesn't like it when there are multiple children inside a Form.Item component (description and input field) as it messes up the form binding. Wrapping the input with its own Form.Item that has the `name` prop fixes this.

BartDeCaluwe committed:
test: update snapshots

BartDeCaluwe committed:
feat: add success and error messages

BartDeCaluwe committed:
feat: add loading message

BartDeCaluwe committed:
feat: swap message box padding values

BartDeCaluwe committed:
feat: display Marketstats in its own grid

BartDeCaluwe committed:
feat: optionally show investment return

BartDeCaluwe committed:
feat: get investment details from the client

BartDeCaluwe committed:
feat: use client to interact with the liquidity pool

BartDeCaluwe committed:
feat: add invest-withdraw page

BartDeCaluwe committed:
test: remove investwithdraw related tests as they require the credix client to work

BartDeCaluwe committed:
feat: handle adding max amount when user has no stake or balance

BartDeCaluwe committed:
Fix: default investment return to 0

BartDeCaluwe committed:
feat: add layout to invest-withdraw page

BartDeCaluwe committed:
fix: message icon szie

BartDeCaluwe committed:
feat: add min/max amount validation

BartDeCaluwe committed:
feat: dont auto hide loading messages

BartDeCaluwe committed:
feat: format numbers

BartDeCaluwe committed:
feat: get market from store

BartDeCaluwe committed:
feat: reload market after transaction

BartDeCaluwe committed:
feat: message is displayed in center of screen and destroyed on close

If we don't destroy the message on close the transparant background remains.

BartDeCaluwe committed:
test: update snapshots

BartDeCaluwe committed:
test: add utils alias

BartDeCaluwe committed:
feat: don't needlessly fetch the market

BartDeCaluwe committed:
feat: remove obsolete files

BartDeCaluwe committed:
feat: fix user base balance seeming to not update when investing or withdrawing

BartDeCaluwe committed:
feat: add step to input

RuvendeGroote commented:
the invest button on the main page doesn't seem to work

RuvendeGroote commented:
![image](https://user-images.githubusercontent.com/14349484/159688766-40cd1079-20da-4969-af54-e79669a18250.png)

RuvendeGroote commented:
![image](https://user-images.githubusercontent.com/14349484/159691068-052814d5-4dc4-4b6c-9ccf-599fc6ab697b.png)

RuvendeGroote commented:
> 

this is just opening the invest withdraw page

BartDeCaluwe committed:
fix: market stats not converting program numbers

BartDeCaluwe committed:
feat: update marketstats copy

BartDeCaluwe committed:
fix: links on index page not navigating

MaximPiessen commented:
<img width="230" alt="Screenshot 2022-03-23 at 14 03 38" src="https://user-images.githubusercontent.com/31801298/159705743-6af8e95c-9daa-470e-8125-e2e11e6dcda9.png">

Typo; but let's change it anyway to "Average financing fee"

MaximPiessen commented:
<img width="771" alt="Screenshot 2022-03-23 at 14 06 12" src="https://user-images.githubusercontent.com/31801298/159706240-95a2e8af-3171-4995-a412-090f5ccadc2f.png">

please use dots for decimals and commas for 1000-separators

MaximPiessen commented:
You can also hide that upward arrow 0% icon for now; we don't have an endpoint yet to populate it with a value.

MaximPiessen commented:
<img width="712" alt="Screenshot 2022-03-23 at 14 11 14" src="https://user-images.githubusercontent.com/31801298/159707155-68f137e7-d8de-430b-8b86-89228736c55c.png">

Make placeholder: "e.g. 100000"

MaximPiessen commented:
<img width="751" alt="Screenshot 2022-03-23 at 14 13 56" src="https://user-images.githubusercontent.com/31801298/159707785-1fd426dd-8d53-4fe1-8115-83e3fec99ff6.png">
align

MaximPiessen commented:
<img width="399" alt="Screenshot 2022-03-23 at 14 15 44" src="https://user-images.githubusercontent.com/31801298/159707979-60a61b0a-2b82-42fd-b5fc-de45ce2e3022.png">
can we also work with the comma's and dots in the notifications to format the numbers?

BartDeCaluwe committed:
feat: RPC optimisations

- change maybeFetchMarket to fetchMarket on invest-withdraw page so we always get a fresh market when the page is loaded
- only fetch the market on invest-withdraw page when a public key is present
- remove a redundant getGatekeepernetwork call
- inline the setting of gatekeepernetwork with a dependency of publicKey and market

BartDeCaluwe committed:
test: update snapshot

BartDeCaluwe committed:
feat: format numbers

BartDeCaluwe committed:
feat: don't show investmentreturn component

BartDeCaluwe committed:
feat: limit investwithdraw component width

BartDeCaluwe committed:
feat: align logo and buttons horizontally

BartDeCaluwe committed:
fix: max withdraw amount not being converted to ui amount

BartDeCaluwe committed:
feat: fix input decimal seperator not being a dot

RuvendeGroote commented on src/message.ts:
```diff
@@ -0,0 +1,53 @@
+}
+
+const defaultClassName = "";
+const defaultOnClose = () => antdMessage.destroy();
```
do we need to destroy it ourselves if we set a duration?

BartDeCaluwe replied:
Normally we shouldn't, no. But when the message is not destroyed, some elements remain in the DOM. With how the design is specified (the dark transparent background) this would cause the background to remain visible.
Destroying the message removes all DOM elements and prevents this from happening.

RuvendeGroote replied:
alright. can you add that as a comment?

BartDeCaluwe replied:
Done!

RuvendeGroote commented on src/message.ts:
```diff
@@ -0,0 +1,53 @@
+const loading = (props: MessageProps) => {
+	return antdMessage.loading({
+		duration: 0,
+		className: defaultClassName,
```
I think we can leave these out as we're not setting it to anything or does it break then?

BartDeCaluwe replied:
The duration you mean? Not setting it to 0 would make it auto hide after 3 seconds.

RuvendeGroote replied:
no, the classname

RuvendeGroote commented on __tests__/validation.test.ts:
```diff
@@ -0,0 +1,57 @@
+import { validateMaxValue, validateMinValue } from "@utils/validation.utils";
+
+describe("validateMaxValue", () => {
```
:thumbsup:

RuvendeGroote commented on src/components/InvestmentDetails.tsx:
```diff
@@ -17,18 +20,23 @@ export const InvestmentDetails = ({
 				<div>
-					<div className="text-2xl font-bold">{`${investments} ${investmentsCurrency}`}</div>
+					<div className="text-2xl font-bold">{`${
+						investments && formatNumber(investments, 2, numberFormatter.format)
```
why the 2 here? the function expects a roundingmode, the enum does have 2 as one of the values but it's less descriptive

RuvendeGroote commented on src/utils/format.utils.ts:
```diff
@@ -4,6 +4,12 @@ import Big, { RoundingMode } from "big.js";
+	maximumFractionDigits: 1,
+});
+
 export const formatNumber = (n: Big, roundingMode: RoundingMode, formatter: any) =>
```
maybe we can just a have a 'round' function that does something like
`(n: Big, roundingMode: RoundingMode, precision?: number) => n.round(precision || 2, roundingMode).toNumber()`
and if we format we can just do something like
`numberFormatter.format(round(principal, RoundingMode.halfup))`

BartDeCaluwe committed:
feat: add rounding util

BartDeCaluwe committed:
test: update snapshot

BouweCeunen commented:
For me when I click the `MAX` amount, it's formatted as such `83941,027925` but I would expect `83,941.027925`. The same formatting as "Current investments", "Current balance" and in the notifications.

BouweCeunen commented on src/components/InvestWithdraw.tsx:
```diff
@@ -1,24 +1,64 @@
+		try {
+			await market.withdraw(toProgramAmount(new Big(amount)));
+			hide();
+			message.success({ content: `Sucessfully withdrew ${formattedNumber} USDC` });
```
`successfully` small typo, also below

BartDeCaluwe committed:
fix: spelling

BartDeCaluwe commented:
> For me when I click the `MAX` amount, it's formatted as such `83941,027925` but I would expect `83,941.027925`. The same formatting as "Current investments", "Current balance" and in the notifications.

Unfortunately, formatting the max amount as `83,941.02...` would mean using the `Intl.numberformatter` which converts it to a string. The `input type="number"`  doesn't accept this. One solution would be to turn the input field into a normal text field, but then you'd lose niceties like a number pad and the step increases.

Also, could you confirm that it is formatted as `83941,027925` with the comma as decimal separator instead of a dot? This shouldn't be the case and was fixed last week. (Also not the case on my machine)

RuvendeGroote commented on src/utils/format.utils.ts:
```diff
@@ -4,13 +4,20 @@ import Big, { RoundingMode } from "big.js";
+export const round = (n: Big, roundingMode: RoundingMode, precision = roundingPrecision) =>
+	n.round(precision, roundingMode);
+
+export const formatNumber = (n: Big, formatter: any) => formatter(n.toNumber());
```
isn't eslint complaining about any?

BartDeCaluwe committed:
fix: add formatter function typing

RuvendeGroote commented:
![image](https://user-images.githubusercontent.com/14349484/160373654-f4eacc0a-3eec-43c0-8d61-0ed6c9afc92c.png)

BouweCeunen commented:
> > For me when I click the `MAX` amount, it's formatted as such `83941,027925` but I would expect `83,941.027925`. The same formatting as "Current investments", "Current balance" and in the notifications.
> 
> Unfortunately, formatting the max amount as `83,941.02...` would mean using the `Intl.numberformatter` which converts it to a string. The `input type="number"` doesn't accept this. One solution would be to turn the input field into a normal text field, but then you'd lose niceties like a number pad and the step increases.
> 
> Also, could you confirm that it is formatted as `83941,027925` with the comma as decimal separator instead of a dot? This shouldn't be the case and was fixed last week. (Also not the case on my machine)

<img width="840" alt="image" src="https://user-images.githubusercontent.com/5576686/160372690-b68c5117-1e40-494a-8a36-769a1754bae8.png">

Weird indeed, I pulled and restarted. Maybe it's related to my system locale? If so, it's ok I guess, it also doesn't happen with Ruven.

BartDeCaluwe committed:
fix: const alias not working

BartDeCaluwe commented:
> Weird indeed, I pulled and restarted. Maybe it's related to my system locale? If so, it's ok I guess, it also doesn't happen with Ruven.

Yeah def, weird. Because mine was also not formatting as it should (due to system locale) and adding `lang="en"` to the input should force the correct formatting (which it did for me and I assume Ruven as well).
